services/ucs/src/ucs/core/controller.py

```diff
@@ -0,0 +1,600 @@
+                    s3_upload_details.s3_upload_id,
+                )
+
+    async def _remove_incomplete_file_upload(
+        self, *, s3_upload_details: S3UploadDetails
+    ) -> None:
+        """Abort an incomplete S3 multipart upload.
+
+        Does not delete any data from the DB.
+
+        Raises:
+        - `UnknownStorageAliasError` if the storage alias is not known.
+        - `UploadAbortError` if there's an error instructing S3 to abort the upload.
+        """
+        file_id = s3_upload_details.file_id
+        storage_alias = s3_upload_details.storage_alias
+        s3_upload_id = s3_upload_details.s3_upload_id
+        bucket_id, object_storage = self._get_bucket_and_storage(
+            storage_alias=storage_alias
+        )
+
+        try:
+            await object_storage.abort_multipart_upload(
+                upload_id=s3_upload_id,
+                bucket_id=bucket_id,
+                object_id=str(file_id),
+            )
+        except object_storage.MultiPartUploadAbortError as err:
+            error = self.UploadAbortError(
+                file_id=file_id, s3_upload_id=s3_upload_id, bucket_id=bucket_id
+            )
+            log.error(
+                error,
+                exc_info=True,
+                extra={
+                    "file_id": file_id,
+                    "bucket_id": bucket_id,
+                    "storage_alias": storage_alias,
+                    "s3_upload_id": s3_upload_id,
+                },
+            )
+            raise error from err
+        except object_storage.MultiPartUploadNotFoundError:
+            # This correspond to an inconsistency between the database and
+            # the storage, however, since this cancel method might be used to
+            # resolve this inconsistency, this exception will be ignored.
+            pass
+
+    async def initiate_file_upload(
+        self, *, box_id: UUID4, alias: str, checksum: str, size: int
+    ) -> UUID4:
+        """Initialize a new multipart upload and return the file ID.
+
+        Raises:
+        - `BoxNotFoundError` if the box does not exist.
+        - `LockedBoxError` if the box exists but is locked.
+        - `FileUploadAlreadyExists` if there's already a FileUpload for this alias.
+        - `UnknownStorageAliasError` if the storage alias is not known.
+        - `OrphanedMultipartUploadError` if an S3 upload is already in progress.
+        """
+        extra: dict[str, Any] = {"box_id": box_id, "alias": alias}
+        # Get the box and create the FileUpload
+        box = await self._get_unlocked_box(box_id=box_id)
+        initiated = now_utc_ms_prec()  # Generate timestamp early to minimize error risk
+        file_id = await self._insert_validated_file_upload(
+            box=box, alias=alias, checksum=checksum, size=size
+        )
+        log.info("FileUpload %s added for alias %s.", file_id, alias, extra=extra)
+
+        # Get the S3 storage details
```

Comment:
Since `_get_bucket_and_storage` can raise an error, you should call it before inserting the file upload to the database.